cmd/syncthing: Refactor cli subcommand to allow flags

imsodin opened the pull request:
Currently the help is quite broken, as kong handles any occurrence of `-h` or `--help`. E.g.  
```
$ ./bin/syncthing cli --home=./config-testing show -h
Usage: syncthing cli [<args> ...]

Command line interface for Syncthing

Arguments:
  [<args> ...]

Flags:
  -h, --help                  Show context-sensitive help.

      --gui-address=URL       Override GUI address (e.g.
                              "http://192.0.2.42:8443")
      --gui-apikey=API-KEY    Override GUI API key
      --home=PATH             Set configuration and data directory
      --conf=PATH             Set configuration directory (config and keys)
```

To fix that it now detects if the `cli` subcommand is used or not before doing any parsing. Thus the entire parsing happens either through kong or urfave/cli, never a mixture (I asked upstream about a "pass-through" mechanism in kong, which does not exist yet and this workaround has been suggested: https://github.com/alecthomas/kong/issues/151).

I removed the db reset operation: I don't think that should be so easily available. I also added a warning to the override operation. Instead I added an `api` subcommand to directly interact with the api, which can be used if an operation isn't provided directly under `show` or `operations`.

AudriusButkevicius commented:
I am pretty sure I added support to handle this in the custom help command handler, so I am surprised this doesn't work.

imsodin commented:
What custom help handler?

I just wanted to split this PR up into the refactor and adding api, but in doing so realized the refactor is also necessary for the api: You can't have flags below `syncthing cli` in the current setup. Kong fails with `syncthing: error: unknown flag --method`.

AudriusButkevicius commented:
I'm referring to this https://github.com/syncthing/syncthing/blob/main/cmd/syncthing/main.go#L252

Which makes this part work:
![image](https://user-images.githubusercontent.com/1144861/110833136-096f6c80-8294-11eb-8e42-9e44862226ac.png)

I guess what doesn't work is:
```syncthing cli subcommand --help```?

AudriusButkevicius commented:
Changing that line I was referring to, to read:
```
	if ctx.Command() == "cli" || ctx.Command() == "cli <args>" {
		return ctx.Run()
	}
```

Fixes the `syncthing cli subcommand --help` not working.

imsodin commented on cmd/syncthing/cli/client.go:
```diff
@@ -63,34 +64,57 @@ func (c *APIClient) Do(req *http.Request) (*http.Response, error) {
 		}
-		body := strings.TrimSpace(string(data))
-		return fmt.Errorf("unexpected HTTP status returned: %s\n%s", response.Status, body)
+		if strings.HasPrefix(body, "404") {
```
Black magic.  
Our httprouter puts the status, `404 resource not found` or similar, into the body. If it's a 404 from us, it's just text (e.g. `no such folder`).

AudriusButkevicius replied:
Definitely comment worthy.

Shouldn't we in that case bark louder, given it's our own CLI with broken routing?

imsodin commented:
That's neat, didn't see that. That can handle the help part. I now digged around `Mapper` and other kong options, and I didn't find any way to pass on arbitrary flags. Any flag that can't be matched generates the `unknown flag` error.

AudriusButkevicius commented:
Can you give me an example to reproduce the issue?

I am not sure I follow what the actual issue is so far, other than something doesn't work.

imsodin commented:
Running e.g. `syncthing cli api --method GET config/folders` fails with `syncthing: error: unknown flag --method`. Any flag that is not defined in a "kong-struct" will produce that error.

AudriusButkevicius commented:
I guess I don't use flags, so it's not a problem, but what you've described looks like a glorified curl, not a CLI.

imsodin commented:
Sure, it is. I kinda guessed you won't be thrilled by it :)  
A major motivation for me is for support: Instead of asking them to get various curl/powershell/whatever options right to give me the api debug info I need, I can just tell them to run `syncthing cli api ...`.  
Also no matter how well we tend the cli, there's likely always stuff that can't be done through it.

imsodin commented:
I've split out the parts that aren't about the refactor to keep these discussions (if any) separate:  
Help: https://github.com/syncthing/syncthing/pull/7483
Removing db reset and other adjustments: https://github.com/syncthing/syncthing/pull/7484

I also removed the `api` command for now. I would still like to have it, however I'd like to focus on the refactor first and it makes the diff a lot smaller:  
I believe the change is necessary, as there's a good chance we will require flags in some `cli` subcommand at some point. E.g. a flag to print pretty or flat json, or a command for events with the various optional arguments there. As far as I see there's no way to do that the way it works right now (urfave/cli command as a kong subcommand).

AudriusButkevicius reviewed (approved):
idk, looks sensible, I assume you've tested it.
